src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets

```diff
@@ -3313,7 +3313,14 @@ because xbuild doesn't support framework reference assemblies.
 </PropertyGroup>
 
 <Target Name="_Deploy">
-  <Exec Command="&quot;$(AdbToolPath)\adb&quot; $(AdbTarget) install -r &quot;$(ApkFileSigned)&quot;" />
+  <PropertyGroup>
+    <_DeployCommand>&quot;$(AdbToolPath)\adb&quot; $(AdbTarget) install -r &quot;$(ApkFileSigned)&quot;</_DeployCommand>
+  </PropertyGroup>
+  <Exec Command="$(_DeployCommand)" ContinueOnError="true" ConsoleToMSBuild="true">
+    <Output TaskParameter="ExitCode" PropertyName="_DeployEditCode" />
+    <Output TaskParameter="ConsoleOutput" PropertyName="_DeployConsoleOutput" />
+  </Exec>
+  <Error Text="The command &quot;$(_DeployCommand)&quot; exitied with code $(_DeployEditCode): $(_DeployConsoleOutput)" Condition="'$(_DeployEditCode)' != '0'" />
```

Comment:
Shouldn't we also/instead generate an XA\d\d\d\d code?  (Note: we'd have to *find* an appropriate code; XA1234 used just for demo purposes.)

Also, should this be a "multi-line" message?

```xml
    <Exec Command="$(_DeployCommand)" ContinueOnError="true" ConsoleToMSBuild="true">
      <Output TaskParameter="ExitCode" PropertyName="_DeployExitCode" />
      <Output TaskParameter="ConsoleOutput" ItemName="_DeployConsoleOutput" />
    </Exec>
    <ItemGroup>
      <_AdbError Include="The command `$(_DeployCommand)` exited with code $(_DeployExitCode):" />
      <_AdbError Include="@(_DeployConsoleOutput->'  %(Identity)')" />
    </ItemGroup>
    <Error
        ContinueOnError="ErrorAndContinue"
        Condition=" '$(_DeployExitCode)' != '0' "
        Code="XA1234"
        Text="%(_AdbError.Identity)"
    />
```

Above would result in a "multi-line" error message, a'la:

```
error XA1234: The command `adb install -r App.apk` exited with code 1:
error XA1234:   line 1 of output
error XA1234:   line 2 of output
error XA1234:   ...
```

...though for my local test (below), the output isn't quite that "clean":

```
Default:
  echo 'this is line 1'; echo 'line 2'; exit 1
  this is line 1
  line 2
/Volumes/Xamarin-Work/xamarin-android/X.targets(8,5): warning MSB3073: The command "echo 'this is line 1'; echo 'line 2'; exit 1" exited with code 1.
  The previous error was converted to a warning because the task was called with ContinueOnError=true.
  Build continuing because "ContinueOnError" on the task "Exec" is set to "true".
/Volumes/Xamarin-Work/xamarin-android/X.targets(16,5): error XA1234: The command `echo 'this is line 1'; echo 'line 2'; exit 1` exited with code 1:
  Build continuing because "ContinueOnError" on the task "Error" is set to "ErrorAndContinue".
/Volumes/Xamarin-Work/xamarin-android/X.targets(16,5): error XA1234:   this is line 1
  Build continuing because "ContinueOnError" on the task "Error" is set to "ErrorAndContinue".
/Volumes/Xamarin-Work/xamarin-android/X.targets(16,5): error XA1234:   line 2
  Build continuing because "ContinueOnError" on the task "Error" is set to "ErrorAndContinue".
Done Building Project "/Volumes/Xamarin-Work/xamarin-android/X.targets" (default targets) -- FAILED.

Build FAILED.

"/Volumes/Xamarin-Work/xamarin-android/X.targets" (default target) (1) ->
(Default target) -> 
  /Volumes/Xamarin-Work/xamarin-android/X.targets(8,5): warning MSB3073: The command "echo 'this is line 1'; echo 'line 2'; exit 1" exited with code 1.


"/Volumes/Xamarin-Work/xamarin-android/X.targets" (default target) (1) ->
(Default target) -> 
  /Volumes/Xamarin-Work/xamarin-android/X.targets(16,5): error XA1234: The command `echo 'this is line 1'; echo 'line 2'; exit 1` exited with code 1:


"/Volumes/Xamarin-Work/xamarin-android/X.targets" (default target) (1) ->
  /Volumes/Xamarin-Work/xamarin-android/X.targets(16,5): error XA1234:   this is line 1


"/Volumes/Xamarin-Work/xamarin-android/X.targets" (default target) (1) ->
  /Volumes/Xamarin-Work/xamarin-android/X.targets(16,5): error XA1234:   line 2

    1 Warning(s)
    3 Error(s)
```

---

Sample test project:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Target Name="Default">
    <PropertyGroup>
      <_Command>echo 'this is line 1'; echo 'line 2'; exit 1</_Command>
      <_DeployCommand>$([MSBuild]::Escape($(_Command)))</_DeployCommand>
    </PropertyGroup>
    <Exec Command="$(_DeployCommand)" ContinueOnError="true" ConsoleToMSBuild="true">
      <Output TaskParameter="ExitCode" PropertyName="_DeployExitCode" />
      <Output TaskParameter="ConsoleOutput" ItemName="_DeployConsoleOutput" />
    </Exec>
    <ItemGroup>
      <_AdbError Include="The command `$(_DeployCommand)` exited with code $(_DeployExitCode):" />
      <_AdbError Include="@(_DeployConsoleOutput->'  %(Identity)')" />
    </ItemGroup>
    <Error
        ContinueOnError="ErrorAndContinue"
        Condition=" '$(_DeployExitCode)' != '0' "
        Code="XA1234"
        Text="%(_AdbError.Identity)"
    />
  </Target>
</Project>
```

Reply:
Error code: have no idea which error code should we use here...
Multiline: why do not use my solution and just print a property `_DeployConsoleOutput`? In that case we will not specify `ContinueOnError="ErrorAndContinue"` and output will be presented better.

Reply:
@jonathanpeppers has the right idea of using an `ADB0000` error code.

Regarding multiline output, with your original property based approach, all lines are merged onto one line using `;` as a separator.  For example, with my "test" target, we'd get:

```
error ADB0000: Errored! this is line 1;line 2
```

If `adb` error output is usually one line, the property approach is fine.  My concern is that if multiple lines are the norm, then "merging" all output onto one line will make it harder to read and understand the output.

Particularly within the IDE, when all the other "noise" generated by MSBuild won't be visible, I think multiline output will be easier to read.